src/libsystemd/sd-event/sd-event.c

```diff
@@ -3935,3 +3965,72 @@ _public_ int sd_event_source_set_floating(sd_event_source *s, int b) {
 
         return 1;
 }
+
+static int ratelimit_timeout_handler(sd_event_source *s, uint64_t usec, void *userdata) {
+        sd_event_source *source = (sd_event_source *) userdata;
+
+        source->ratelimit->effective = false;
+
+        (void) sd_event_source_set_enabled(source, SD_EVENT_ON);
+        (void) sd_event_source_set_enabled(s, SD_EVENT_OFF);
+
+        return 1;
+}
+
+_public_ int sd_event_source_set_ratelimit(sd_event_source *s, uint64_t interval, uint64_t burst) {
+        int r;
+
+        assert_return(s, -EINVAL);
+
+        r = sd_event_source_disable_ratelimit(s);
+        if (r < 0)
+                return r;
+
+        s->ratelimit = new0(EventSourceRateLimit, 1);
+        if (!s->ratelimit)
+                return -ENOMEM;
+
+        s->ratelimit->r = (RateLimit) { .interval = interval * USEC_PER_SEC, .burst = burst };
+
+        r = sd_event_add_time(s->event, &s->ratelimit->timeout_source, clock_boottime_or_monotonic(), interval * USEC_PER_SEC / burst, 0, ratelimit_timeout_handler, s);
+        if (r < 0)
+                return r;
+
+        r = sd_event_source_set_enabled(s->ratelimit->timeout_source, SD_EVENT_OFF);
+        if (r < 0)
+                return r;
+
+        return 1;
+}
+
+_public_ int sd_event_source_get_ratelimit(sd_event_source *s, uint64_t *interval, uint64_t *burst) {
+        assert_return(s && interval && burst, -EINVAL);
+
+        if (!s->ratelimit)
+                return 0;
+
+        *interval = s->ratelimit->r.interval / USEC_PER_SEC;
+        *burst = s->ratelimit->r.burst;
+
+        return 1;
+}
+
+_public_ int sd_event_source_disable_ratelimit(sd_event_source *s) {
```

Comment:
hmm, why an extra API call for this, shouldm't sd_event_source_set_ratelimit(s, 0, 0) be good enough for that?